File: data-structures/non-linear/binary-tree/BinaryTree.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Node {
public:
  int data;
  Node *left;
  Node *right;

  Node(int value) {
    data = value;
    left = right = nullptr;
  }
};

static int index = -1;

Node *buildBT(vector<int> preorder) {
  index++;

  if (preorder[index] == -1) {
    return nullptr;
  }

  Node *root = new Node(preorder[index]);
  root->left = buildBT(preorder);
  root->right = buildBT(preorder);

  return root;
}

void preorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  cout << root->data << " ";
  preorderTraversal(root->left);
  preorderTraversal(root->right);
}

void inorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  inorderTraversal(root->left);
  cout << root->data << " ";
  inorderTraversal(root->right);
}

void postorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  postorderTraversal(root->left);
  postorderTraversal(root->right);
  cout << root->data << " ";
}

void levelorderTraversal(Node *root) {
  queue<Node *> q;

  q.push(root);
  q.push(nullptr);

  while (q.size() > 0) {
    Node *temp = q.front();

    q.pop();

    if (temp == nullptr) {
      if (!q.empty()) {
        cout << "\n";
        q.push(nullptr);
        continue;
      } else {
        break;
      }
    }

    cout << temp->data << " ";

    if (temp->left != nullptr) {
      q.push(temp->left);
    }

    if (temp->right != nullptr) {
      q.push(temp->right);
    }
  }

  cout << "\n";
}

File: data-structures/non-linear/trees/binary-tree/iterative/BinaryTree.ts
import type { TreeNode } from "../../node/TreeNode";

function preOrderTraversal(root: TreeNode | null): void {
	if (root == null) return;

	const stack: Array<TreeNode> = [];
	let tempNode: TreeNode;
	stack.push(root);

	while (stack.length !== 0) {
		//@ts-expect-error empty array pop returns undefined but that is already checked in the while loop
		tempNode = stack.pop();

		console.log(`${tempNode.getValue()}`);
		if (tempNode.getRight() != null) {
			//@ts-expect-error checking has been done for right being null
			stack.push(tempNode.getRight());
		}

		if (tempNode.getLeft() != null) {
			//@ts-expect-error
			stack.push(tempNode.getRight());
		}
	}
}

function inOrderTraversal(root: TreeNode | null): void {
	if (root == null) return;

	const stack: Array<TreeNode> = [];
	let node: TreeNode | null = root;

	while (node != null || stack.length !== 0) {
		if (node != null) {
			stack.push(node);
			node = node.getLeft();
		} else {
			//@ts-expect-error
			node = stack.pop();
			console.log(`${node?.getValue()}`);
			//@ts-expect-error
			node = node.getRight();
		}
	}
}

function postOrderTraversal(root: TreeNode | null): void {
	const stack: Array<TreeNode> = [];

	let lastVisitedNode: TreeNode | null = null;
	let node: TreeNode | null = root;

	while (node != null || stack.length !== 0) {
		if (node != null) {
			stack.push(node);
			node = node.getLeft();
		} else {
			const top: TreeNode | null = stack[-1];

			if (top.getRight() != null && lastVisitedNode !== top.getRight()) {
				node = top.getRight();
			} else {
				console.log(`${top.getValue()}`);
				lastVisitedNode = top;
				stack.pop();
			}
		}
	}
}

export { inOrderTraversal, postOrderTraversal, preOrderTraversal };
